io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -808,11 +809,15 @@ schedule thread@Thread{
           simstate' = simstate { clocks = Map.insert clockid' clockoff clocks }
       schedule thread' simstate'
 
+    NewTimeout d k | d < 0 -> do
+      let t       = Timeout Nothing nextTmid
+          thread' = thread { threadControl = ThreadControl (k t) ctl }
+      schedule thread' simstate { nextTmid = succ nextTmid }
     NewTimeout d k -> do
       tvar <- execNewTVar nextVid TimeoutPending
       let expiry  = d `addTime` time
-          t       = Timeout tvar nextTmid
-          timers' = PSQ.insert nextTmid expiry tvar timers
+          t       = Timeout (Just tvar) nextTmid
+          timers' = if d >= 0 then PSQ.insert nextTmid expiry tvar timers else timers
```

Comment:
Why do we need to test `d` here? We already know `d >= 0` because we handled the `NewTimeout d k | d < 0` case above.

Reply:
I just haven't reviewed it myself ... this should not be here.